### ISSUE TITLE
[azservicebus] panic in ReceiveMessages on context timeout (invalid memory address or nil pointer dereference)

### ISSUE THREAD

**Issue by peterzeller** [Azure/azure-sdk-for-go#16923]
### Bug Report

- import path of package in question: ``"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"`
- SDK version (from Go.mod): 

```
        github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus v0.3.4-0.20220126013851-916bcb2ef21f
	github.com/Azure/azure-sdk-for-go/sdk/azcore v0.21.0 // indirect
	github.com/Azure/azure-sdk-for-go/sdk/internal v0.8.3 // indirect
	github.com/Azure/azure-sdk-for-go/sdk/messaging/internal v0.0.0-20211208010914-2b10e91d237e // indirect
	github.com/Azure/go-amqp v0.17.0 // indirect
```

- output of `go version`: `go version go1.17.3 linux/amd64`

- What happened?

I call `ReceiveMessages(ctx, 95, &azservicebus.ReceiveMessagesOptions{})` and get the following panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x712de5]
goroutine 199 [running]:
github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).receiveMessagesImpl(0xc000138150, {0x902a68, 0xc000364720}, 0x64, 0xc00038bc10)
	/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/messaging/azservicebus@v0.3.4-0.20220125201221-b3f0e1078c77/receiver.go:432 +0x265
github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).ReceiveMessages(0xc000138150, {0x902a68, 0xc000364720}, 0xb1f736cc3, 0xbba120)
	/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/messaging/azservicebus@v0.3.4-0.20220125201221-b3f0e1078c77/receiver.go:184 +0x1ea
```

- How can we reproduce it?

The error seems to happen when the context times out. I tried to get a reproducible test with the code below, but it does not happen every run.

```
import (
	"context"
	"errors"
	"fmt"
	"math/rand"
	"os"
	"sync"
	"testing"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/admin"
	"github.com/stretchr/testify/require"
)

func TestAzureContextBug(t *testing.T) {
	connectionStr := os.Getenv("ASB_CONNECTION_STRING")
	require.NotEmpty(t, connectionStr, "Define ASB_CONNECTION_STRING environment variable to run this test")

	adminClient, err := admin.NewClientFromConnectionString(connectionStr, &admin.ClientOptions{})
	require.NoError(t, err)

	rand.Seed(time.Now().UnixNano())
	topicName := fmt.Sprintf("test-topic-%d", rand.Intn(1000))
	subscriptionName := "test-subscription"

	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()

	autoDeleteOnIdle := 5 * time.Minute
	_, err = adminClient.CreateTopic(ctx, topicName, &admin.TopicProperties{
		AutoDeleteOnIdle: &autoDeleteOnIdle,
	}, &admin.CreateTopicOptions{})
	require.NoError(t, err)

	t.Cleanup(func() {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		_, err := adminClient.DeleteTopic(ctx, topicName, &admin.DeleteTopicOptions{})
		require.NoError(t, err)
	})

	_, err = adminClient.CreateSubscription(ctx, topicName, subscriptionName, &admin.SubscriptionProperties{}, &admin.CreateSubscriptionOptions{})
	require.NoError(t, err)



	// publish messages


	numMessages := 1000
	batchSize := 100

	publisher := func() error {
		client, err := azservicebus.NewClientFromConnectionString(connectionStr, &azservicebus.ClientOptions{})
		if err != nil {
			return err
		}

		sender, err := client.NewSender(topicName, &azservicebus.NewSenderOptions{})
		if err != nil {
			return err
		}
		for i := 0; i < numMessages; i += batchSize {
			batch, err := sender.NewMessageBatch(ctx, &azservicebus.MessageBatchOptions{})
			if err != nil {
				return err
			}
			for j := 0; j < batchSize; j++ {
				msg := fmt.Sprintf("message-%d", i+j)
				err := batch.AddMessage(&azservicebus.Message{
					Body: []byte(msg),
				})
				if err != nil {
					return err
				}
			}
			err = sender.SendMessageBatch(ctx, batch)
			if err != nil {
				return err
			}
			//time.Sleep(10*time.Millisecond)
			t.Logf("published %d/%d messages", i+batchSize, numMessages)
		}
		t.Logf("publisher finished")
		return nil
	}
	errorChan := make(chan error)
	go func() {
		errorChan <- publisher()
	}()


	// receive messages
	client, err := azservicebus.NewClientFromConnectionString(connectionStr, &azservicebus.ClientOptions{})
	require.NoError(t, err)

	receiver, err := client.NewReceiverForSubscription(topicName, subscriptionName, &azservicebus.ReceiverOptions{})
	require.NoError(t, err)

	receivedCount := 0
	waitTime := 0 * time.Nanosecond
	for receivedCount < numMessages && ctx.Err() == nil {
		//ctx2 := newCustomContext(waitTime)
		ctx2, cancel2 := context.WithTimeout(context.Background(), waitTime)
		defer cancel2()

		t.Logf("waitTime = %v", waitTime)
		messages, err := receiver.ReceiveMessages(ctx2, 95, &azservicebus.ReceiveMessagesOptions{})
		if errors.Is(err, context.DeadlineExceeded) {
			t.Logf("context deadline exceeded")
		} else {
			require.NoError(t, err)
		}

		if len(messages) == 0 {
			continue
		}
		receivedCount += len(messages)
		t.Logf("Message count: %d (total: %d)", len(messages), receivedCount)
		waitTime = waitTime + 100*time.Nanosecond

		var wg sync.WaitGroup
		for _, message := range messages {
			body, err := message.Body()
			require.NoError(t, err)
			require.NotNil(t, body)
			//t.Logf("received %d: %s", j, body)
			wg.Add(1)
			go func(message *azservicebus.ReceivedMessage) {
				defer wg.Done()
				err = receiver.CompleteMessage(ctx, message)
				require.NoError(t, err)
			}(message)
		}
		wg.Wait()
	}
	require.Equal(t, numMessages, receivedCount)
	require.NoError(t, <- errorChan, "error in publisher")
}

```

**Edit**: Maybe also related: In other runs I got a deadlock in the same line. This is the stack trace after the test was killed after 10 minutes:

```
goroutine 1881 [select, 8 minutes]:
github.com/Azure/go-amqp.(*manualCreditor).Drain(0xc0007e8150, {0x902a30, 0xc000136000}, 0xc00038fa40)
	/go/pkg/mod/github.com/!azure/go-amqp@v0.17.0/manualCreditor.go:79 +0x190
github.com/Azure/go-amqp.(*link).DrainCredit(0xc00038fa40, {0x902a30, 0xc000136000})
	/go/pkg/mod/github.com/!azure/go-amqp@v0.17.0/link.go:607 +0x77
github.com/Azure/go-amqp.(*Receiver).DrainCredit(0x8f9ec0, {0x902a30, 0xc000136000})
	/go/pkg/mod/github.com/!azure/go-amqp@v0.17.0/receiver.go:38 +0x26
github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).receiveMessagesImpl(0xc0001e8540, {0x902a68, 0xc000aac180}, 0x64, 0xc000c1d4e0)
	/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/messaging/azservicebus@v0.3.4-0.20220126013851-916bcb2ef21f/receiver.go:432 +0x282
github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).ReceiveMessages(0xc0001e8540, {0x902a68, 0xc000aac180}, 0x1, 0x11)
	/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/messaging/azservicebus@v0.3.4-0.20220126013851-916bcb2ef21f/receiver.go:184 +0x1ea
```

**Comment by richardpark-msft** [comment]
Hi @peterzeller,

I've got a PR out for the panic (#16941). 

The deadlock is interesting - we did have a fix in go-amqp v0.17 for a potential deadlock around drain. I'll try out your test case and see if I can reproduce it - we had some discussions about how to make that bit more robust.

**Comment by richardpark-msft** [comment]
Hi @peterzeller, 0.3.4 has been released and will fix the panic issue you mentioned.

The drain issue is still being worked on, but I hope to have something within this week.
